// mobx-angular/esm5/lib/mobx-autorun.directive.js
import { __decorate } from "tslib";
import { Directive, ViewContainerRef, TemplateRef, OnInit, OnDestroy, Input, EmbeddedViewRef } from '@angular/core';
import { autorun } from 'mobx';
// import { mobxAngularDebug } from '../utils/mobx-angular-debug';
var MobxAutorunDirective = /** @class */ (function () {
    function MobxAutorunDirective(templateRef, viewContainer) {
        this.templateRef = templateRef;
        this.viewContainer = viewContainer;
        this.templateBindings = {};
    }
    MobxAutorunDirective.prototype.ngOnInit = function () {
        this.view = this.viewContainer.createEmbeddedView(this.templateRef);
        if (this.dispose) {
            this.dispose();
        }
        if (this.shouldDetach()) {
            this.view.detach();
        }
        this.autoDetect(this.view);
        // mobxAngularDebug(this.view, this.dispose);
    };
    MobxAutorunDirective.prototype.shouldDetach = function () {
        return this.mobxAutorun && this.mobxAutorun.detach;
    };
    MobxAutorunDirective.prototype.autoDetect = function (view) {
        this.dispose = autorun(function () { return view.detectChanges(); });
    };
    MobxAutorunDirective.prototype.ngOnDestroy = function () {
        if (this.dispose) {
            this.dispose();
        }
    };
    MobxAutorunDirective.ctorParameters = function () { return [
        { type: TemplateRef },
        { type: ViewContainerRef }
    ]; };
    __decorate([
        Input()
    ], MobxAutorunDirective.prototype, "mobxAutorun", void 0);
    MobxAutorunDirective = __decorate([
        Directive({ selector: '[mobxAutorun]' })
    ], MobxAutorunDirective);
    return MobxAutorunDirective;
}());
export { MobxAutorunDirective };
