// mobx-angular/esm5/public-api.js
/*
 * Public API Surface of mobx-angular
 */
import { __read, __spread } from "tslib";
import { action as mobxAction } from 'mobx';
import { computed as mobxComputed } from 'mobx';
import { observable as mobxObservable } from 'mobx';
export * from './lib/mobx-autorun.directive';
export * from './lib/mobx-reaction.directive';
export * from './lib/mobx-angular.module';
// Re-export mobx operators to be able to use inside components with AOT:
export function actionInternal() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    return mobxAction.apply(void 0, __spread(args));
}
export var action = Object.assign(actionInternal, mobxAction);
function computedInternal() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    return mobxComputed.apply(void 0, __spread(args));
}
export var computed = Object.assign(computedInternal, mobxComputed);
function observableInternal() {
    var args = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        args[_i] = arguments[_i];
    }
    return mobxObservable.apply(void 0, __spread(args));
}
export var observable = Object.assign(observableInternal, mobxObservable);
